File: python/day_04/main.py
from __future__ import annotations
from dataclasses import dataclass


@dataclass
class Range:
    lower_bound: int
    upper_bound: int

    def contains(self, range: Range) -> int:
        condition_1 = self.lower_bound <= range.lower_bound
        condition_2 = range.upper_bound <= self.upper_bound
        return condition_1 and condition_2

    def overlaps(self, range: Range) -> int:
        condition_1 = self.lower_bound <= range.upper_bound
        condition_2 = range.lower_bound <= self.upper_bound
        return condition_1 and condition_2


def line_to_ranges(input_line: str) -> tuple[Range, Range]:

    raw_range_1, raw_range_2 = input_line.split(",")

    lower_bound_1, upper_bound_1 = raw_range_1.split("-")
    lower_bound_2, upper_bound_2 = raw_range_2.split("-")

    range_1 = Range(int(lower_bound_1), int(upper_bound_1))
    range_2 = Range(int(lower_bound_2), int(upper_bound_2))

    return range_1, range_2


def contained(range_1: Range, range_2: Range) -> int:
    return range_1.contains(range_2) or range_2.contains(range_1)


def main(input_data: str) -> tuple[int, int]:

    range_pairs = [line_to_ranges(input_line) for input_line in input_data.splitlines()]

    part_1_answer = sum(contained(range_1, range_2) for range_1, range_2 in range_pairs)
    part_2_answer = sum(range_1.overlaps(range_2) for range_1, range_2 in range_pairs)

    return part_1_answer, part_2_answer

File: rust/src/day_04/main.rs
struct Range {
    lower_bound: i32,
    upper_bound: i32,
}

impl Range {
    fn new(lower_bound: i32, upper_bound: i32) -> Self {
        Self {
            lower_bound: lower_bound,
            upper_bound: upper_bound,
        }
    }

    fn contains(&self, range: &Range) -> bool {
        let condition_1 = self.lower_bound <= range.lower_bound;
        let condition_2 = range.upper_bound <= self.upper_bound;
        condition_1 && condition_2
    }

    fn overlaps(&self, range: &Range) -> bool {
        let condition_1 = self.lower_bound <= range.upper_bound;
        let condition_2 = range.lower_bound <= self.upper_bound;
        condition_1 && condition_2
    }
}

fn line_to_ranges(input_line: &str) -> (Range, Range) {
    let raw_range = input_line.split(",").collect::<Vec<_>>();

    assert_eq!(raw_range.len(), 2);
    let (raw_range_1, raw_range_2) = (raw_range[0], raw_range[1]);

    let bounds_1 = raw_range_1
        .split("-")
        .map(|bound| bound.parse::<i32>().unwrap())
        .collect::<Vec<_>>();

    assert_eq!(bounds_1.len(), 2);
    let (lower_bound_1, upper_bound_1) = (bounds_1[0], bounds_1[1]);

    let bounds_2 = raw_range_2
        .split("-")
        .map(|bound| bound.parse::<i32>().unwrap())
        .collect::<Vec<_>>();

    assert_eq!(bounds_2.len(), 2);
    let (lower_bound_2, upper_bound_2) = (bounds_2[0], bounds_2[1]);

    let range_1 = Range::new(lower_bound_1, upper_bound_1);
    let range_2 = Range::new(lower_bound_2, upper_bound_2);

    (range_1, range_2)
}

fn contained(range_1: &Range, range_2: &Range) -> bool {
    range_1.contains(range_2) || range_2.contains(range_1)
}

pub fn main(input_data: String) -> (i32, i32) {
    let range_pairs = input_data
        .lines()
        .map(|input_line| line_to_ranges(input_line));

    let part_1_answer = range_pairs
        .clone()
        .map(|(range_1, range_2)| contained(&range_1, &range_2) as i32)
        .sum();

    let part_2_answer = range_pairs
        .map(|(range_1, range_2)| range_1.overlaps(&range_2) as i32)
        .sum();

    (part_1_answer, part_2_answer)
}
